crates/bevy_app/src/app.rs

```diff
@@ -732,144 +571,128 @@ impl App {
     ) -> &mut Self {
-        let registry = self.world.resource_mut::<AppTypeRegistry>();
-        registry.write().register_type_data::<T, D>();
+        self.main_mut().register_type_data::<T, D>();
         self
     }
 
-    /// Retrieves a `SubApp` stored inside this [`App`].
+    /// Returns a reference to the [`World`].
+    pub fn world(&self) -> &World {
+        self.main().world()
+    }
+
+    /// Returns a mutable reference to the [`World`].
+    pub fn world_mut(&mut self) -> &mut World {
+        self.main_mut().world_mut()
+    }
+
+    /// Returns a reference to the main [`SubApp`].
+    pub fn main(&self) -> &SubApp {
+        &self.sub_apps.main
+    }
+
+    /// Returns a mutable reference to the main [`SubApp`].
+    pub fn main_mut(&mut self) -> &mut SubApp {
+        &mut self.sub_apps.main
+    }
+
+    /// Returns a reference to the [`SubApp`] with the given label.
     ///
     /// # Panics
     ///
-    /// Panics if the `SubApp` doesn't exist.
-    pub fn sub_app_mut(&mut self, label: impl AppLabel) -> &mut App {
-        match self.get_sub_app_mut(label) {
-            Ok(app) => app,
-            Err(label) => panic!("Sub-App with label '{:?}' does not exist", label.as_str()),
-        }
-    }
-
-    /// Retrieves a `SubApp` inside this [`App`] with the given label, if it exists. Otherwise returns
-    /// an [`Err`] containing the given label.
-    pub fn get_sub_app_mut(&mut self, label: impl AppLabel) -> Result<&mut App, AppLabelId> {
-        let label = label.as_label();
-        self.sub_apps
-            .get_mut(&label)
-            .map(|sub_app| &mut sub_app.app)
-            .ok_or(label)
+    /// Panics if the sub-app doesn't exist.
+    pub fn sub_app(&self, label: &dyn AppLabel) -> &SubApp {
+        self.get_sub_app(label).unwrap_or_else(|| {
+            panic!("No sub-app with label '{:?}' exists.", label.as_str());
+        })
     }
 
-    /// Retrieves a `SubApp` stored inside this [`App`].
+    /// Returns a reference to the [`SubApp`] with the given label.
     ///
     /// # Panics
     ///
-    /// Panics if the `SubApp` doesn't exist.
-    pub fn sub_app(&self, label: impl AppLabel) -> &App {
-        match self.get_sub_app(label) {
-            Ok(app) => app,
-            Err(label) => panic!("Sub-App with label '{:?}' does not exist", label.as_str()),
-        }
+    /// Panics if the reference to a sub-app doesn't exist.
+    pub fn sub_app_mut(&mut self, label: &dyn AppLabel) -> &mut SubApp {
+        self.get_sub_app_mut(label).unwrap_or_else(|| {
+            panic!("No sub-app with label '{:?}' exists.", label.as_str());
+        })
     }
 
-    /// Inserts an existing sub app into the app
-    pub fn insert_sub_app(&mut self, label: impl AppLabel, sub_app: SubApp) {
-        self.sub_apps.insert(label.as_label(), sub_app);
+    /// Returns a reference to the [`SubApp`] with the given label, if it exists.
+    pub fn get_sub_app(&self, label: &dyn AppLabel) -> Option<&SubApp> {
```

Comment:
The migration guide should mention that this function now returns an `Option<SubApp>` instead of a `Result<App>`